src/sys/prctl.rs

```diff
@@ -0,0 +1,229 @@
+
+    #[repr(i32)]
+    #[non_exhaustive]
+    #[allow(non_camel_case_types)]
+    pub enum PrctlMCEKillPolicy {
+        /// The thread will receive SIGBUS as soon as a memory corruption is detected.
+        PR_MCE_KILL_EARLY,
+        /// The process is killed only when it accesses a corrupted page.
+        PR_MCE_KILL_LATE,
+        /// Uses the system-wide default.
+        PR_MCE_KILL_DEFAULT,
+    }
+    impl TryFrom<i32>
+}
+
+fn prctl_set_bool(option: PrctlOption, status: bool) -> Result<()> {
+    let res = unsafe { libc::prctl(option as c_int, status as c_ulong, 0, 0, 0) };
+    Errno::result(res).map(drop)
+}
+
+/// Set the "child subreaper" attribute for this process
+pub fn set_child_subreaper(attribute: bool) -> Result<()> {
+    prctl_set_bool(PR_SET_CHILD_SUBREAPER, attribute)
+}
+
+/// Get the "child subreaper" attribute for this process
+pub fn get_child_subreaper() -> Result<bool> {
+    // prctl writes into this var
+    let mut subreaper: c_int = 0;
+    let subreaper_ptr: *mut c_int = &mut subreaper;
+
+    let res = unsafe {
+        libc::prctl(PR_GET_CHILD_SUBREAPER as c_int, subreaper_ptr as c_ulong, 0, 0, 0)
+    };
+
+    Errno::result(res).map(|_| subreaper != 0)
+}
+
+/// Set the dumpable attribute which determines if core dumps are created for this process.
+pub fn set_dumpable(attribute: bool) -> Result<()> {
+    prctl_set_bool(PR_SET_DUMPABLE, attribute)
+}
+
+/// Get the dumpable attribute for this process.
+pub fn get_dumpable() -> Result<bool> {
+    let res = unsafe {
+        libc::prctl(PR_GET_DUMPABLE as c_int, 0, 0, 0, 0)
+    };
+
+    Errno::result(res).map(|res| res != 0)
+}
+
+/// Set the "keep capabilities" attribute for this process. This causes the thread to retain
+/// capabilities even if it switches its UID to a nonzero value.
+pub fn set_keepcaps(attribute: bool) -> Result<()> {
+    prctl_set_bool(PR_SET_KEEPCAPS, attribute)
+}
+
+/// Get the "keep capabilities" attribute for this process
+pub fn get_keepcaps() -> Result<bool> {
+    let res = unsafe {
+        libc::prctl(PR_GET_KEEPCAPS as c_int, 0, 0, 0, 0)
+    };
+
+    Errno::result(res).map(|res| res != 0)
+}
+
+/// Clear the thread memory corruption kill policy and use the system-wide default
+pub fn clear_mce_kill() -> Result<()> {
+    let res = unsafe {
+        libc::prctl(PR_MCE_KILL as c_int, libc::PR_MCE_KILL_CLEAR, 0, 0, 0)
+    };
+
+    Errno::result(res).map(drop)
+}
+
+/// Set the thread memory corruption kill policy
+pub fn set_mce_kill(policy: PrctlMCEKillPolicy) -> Result<()> {
+    let res = unsafe {
+        libc::prctl(PR_MCE_KILL as c_int, libc::PR_MCE_KILL_SET, policy as c_ulong, 0, 0)
+    };
+
+    Errno::result(res).map(drop)
+}
+
+/// Get the thread memory corruption kill policy
+pub fn get_mce_kill() -> Result<PrctlMCEKillPolicy> {
+    let res = unsafe {
+        libc::prctl(PR_MCE_KILL_GET as c_int, 0, 0, 0, 0)
+    };
+
+    Errno::result(res).map(|res| PrctlMCEKillPolicy::try_from(res).unwrap())
```

Comment:
This unwrap could panic in the future if the kernel adds a new kill policy.  Better to return an error in that case.

Reply:
Fixed it :+1: